server/src/main/java/org/elasticsearch/cluster/routing/IndexShardRoutingTable.java

```diff
@@ -663,6 +664,16 @@ static boolean noAssignedReplicaWithoutActivePrimary(List<ShardRouting> shards)
             return seenAssignedReplica == false;
         }
 
+        static boolean noRelocatingUnsearchableShards(List<ShardRouting> shards) {
+            // this is unsupported until ES-4677 is implemented
+            for (var shard : shards) {
+                if (shard.role() == ShardRouting.Role.INDEX_ONLY && shard.state() == ShardRoutingState.RELOCATING) {
+                    return false;
+                }
```

Comment:
Also maybe just assert here so we can report the details:

```suggestion
                assert shard.isSearchable() || shard.state() != ShardRoutingState.RELOCATING : shard;
```